core/src/main/java/org/jruby/RubySymbol.java

```diff
@@ -1430,6 +1430,11 @@ public static int javaStringHashCode(ByteList iso8859) {
         return h;
     }
 
+    @Override
+    public Class getJavaClass() {
```

Comment:
considered this in the past but always backed-off since it won't auto-convert on a round-trip and IMHO leads to more "inconsistency".

e.g. having a Ruby string key in a hash can be looked up with a Java string and also a Java string value gets converted to RubyString. we can't have the same parity with a symbol so it seemed better to just fail than eventually end up w RubySymbol -> java.lang.String -> RubyString

Reply:
I think the fix for https://github.com/jruby/jruby/issues/8015 does not need this change, or does it?

if it does we should add more tests covering the changed behavior as the (default) Ruby -> Java (`JavaUtil.convertRubyToJava`) mechanism will end up converting `RubySymbol` -> `java.lang.String` e.g. when seeing a generic `java.lang.Object` (param) type or when doing a generic `call_from_java_a_method_returning_symbol`.

the current behavior is that it passed down `org.jruby.RubySymbol` as is, this could be considered a bug as well as a breaking change (I am fine to consider it a bug atm but more tests on what is changing could reveal concerns).

sorry I haven't looked in detail, hopefully my comments are somehow useful :wink:

Reply:
Thank you for taking the time to review this. It will take me some time to dig around and digest everything as this is completely new territory for me.

Actually, I think the other change included here is not necessary for this fix. That code path is only hit when the resolution is ambiguous (ambiguous = true in CallableSelector.findMatchingCallableForArgs).

I will have a go at making the other suggested change in #8015 if that is preferred. It seems that may have fewer unintentional side effects.

Reply:
You were on the spot with the changes, it's just that the JI coverage isn't that good opposed to what's out there in terms of usage. I would have been fine shipping the previous version if we added some more tests, as well.

The new approach seems fine with me, although probably worth revisit later as that special case handling seems a bit "redundant".

We can revisit the approach in another PR with the test included... all in all :+1: job!

Reply:
Wow just a passing comment but it sucks that to generalize all these arities it would force boxing the arguments.  I do see that boxing occurring in deeper calls but generally only on first hit to find an invoker.  I am curious if there is some way of moving logic back into findCallAble* and then reintroducing static methods with more attention to not passing multiple arity signatures?  (Note: definitely not in this PR just for future work).  Not sure if that would lead to simpler/smaller or complicated/larger but it looks like an enticing puzzle.

Reply:
@kares ok, I am going back to the original implementation. Could you let me know what sort of tests you are looking for, and I will take a crack at them. Am I on the right track with what I have written? Do you want to see something more direct (eg. writing some unit tests for CallableSelector in `core/src/test/java/org/jruby/java/dispatch/CallableSelectorTest.java`?)

Thanks again & sorry for the churn

Reply:
actually, I am sorry, apparently the `getJavaClass` is not sufficient to cause the behavior I assumed:

```ruby
pp ({ 'foo' => :bar }).to_java.get('foo')
```
I expected your changes to cause this to print a `"bar"` string instead of symbol (current behavior).

... for the record, the test I had in mind would have been smt like:
```diff
diff --git spec/java_integration/types/coercion_spec.rb spec/java_integration/types/coercion_spec.rb
index c15c0df903..bee9e30d6c 100644
--- spec/java_integration/types/coercion_spec.rb
+++ spec/java_integration/types/coercion_spec.rb
@@ -9,7 +9,7 @@ java_import "java_integration.fixtures.PackageConstructor"
 java_import "java_integration.fixtures.ProtectedConstructor"
 java_import "java_integration.fixtures.PrivateConstructor"
 
-describe "Java String and primitive-typed methods" do
+describe "Java coercion" do
   it "should coerce to Ruby types when returned" do
     expect(CoreTypeMethods.getString).to be_kind_of(String)
     expect(CoreTypeMethods.getString).to eq("foo")
@@ -207,6 +207,15 @@ describe "Java Object-typed methods" do
     expect(CoreTypeMethods.getObjectType(1 << 128)).to eq("class java.math.BigInteger")
   end
 
+  it "converts symbols to Java string" do
+    expect(CoreTypeMethods.getObjectType(:foo)).to eq("class java.lang.String")
+  end
+
   it "passes coerced to_java values - keeping the Java type" do
     expect(CoreTypeMethods.getObjectType('foo'.to_java)).to eq "class java.lang.String"
     expect(CoreTypeMethods.getObjectType(0.to_java)).to eq "class java.lang.Long"
```
current behavior is: `"class org.jruby.RubySymbol"`  and that behavior does not change with your changes in the PR. sorry again for the confusion...